src/Symfony/Component/Security/Http/Authenticator/FormLoginAuthenticator.php

```diff
@@ -123,16 +123,16 @@ private function getCredentials(Request $request): array
             $credentials['password'] = ParameterBagUtils::getRequestParameterValue($request, $this->options['password_parameter']) ?? '';
         }
 
-        if (!\is_string($credentials['username']) && !$credentials['username'] instanceof \Stringable) {
-            throw new BadRequestHttpException(sprintf('The key "%s" must be a string, "%s" given.', $this->options['username_parameter'], \gettype($credentials['username'])));
+        if (!(\is_string($credentials['username']) || $credentials['username'] instanceof \Stringable) || '' === $credentials['username']) {
+            throw new BadRequestHttpException(sprintf('The key "%s" must be a non-empty string, "%s" given.', $this->options['username_parameter'], \gettype($credentials['username'])));
         }
 
         $credentials['username'] = trim($credentials['username']);
 
         $request->getSession()->set(SecurityRequestAttributes::LAST_USERNAME, $credentials['username']);
 
-        if (!\is_string($credentials['password']) && (!\is_object($credentials['password']) || !method_exists($credentials['password'], '__toString'))) {
-            throw new BadRequestHttpException(sprintf('The key "%s" must be a string, "%s" given.', $this->options['password_parameter'], \gettype($credentials['password'])));
+        if (!(\is_string($credentials['password']) || $credentials['password'] instanceof \Stringable) || '' === $credentials['password']) {
+            throw new BadRequestHttpException(sprintf('The key "%s" must be a non-empty string, "%s" given.', $this->options['password_parameter'], \gettype($credentials['password'])));
```

Comment:
I would separate the check of the type (where using gettype in the message makes sense) from the check for the emptyness (which might deserve a different exception message)